misc/python/materialize/output_consistency/data_generation/generator.py

```diff
@@ -0,0 +1,102 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from materialize.output_consistency.expressions.expression import (
+    Expression,
+    ExpressionWithNArgs,
+)
+from materialize.output_consistency.operations.operation import OperationWithNArgs
+from materialize.output_consistency.operations.operation_provider import OPERATION_TYPES
+from materialize.output_consistency.values.data_provider import DATA_TYPES
+from materialize.output_consistency.values.data_type import RawValue
+
+
+def create_expressions() -> list[ExpressionWithNArgs]:
+    expressions = []
+
+    for data_type in DATA_TYPES:
+
+        if len(data_type.raw_values) == 0:
+            continue
+
+        for operation_type in OPERATION_TYPES:
+            combinations = get_combinations(
+                data_type.raw_values,
+                offset_value=data_type.raw_values[0],
+                length=operation_type.op_count,
+                with_self=True,
+                with_earlier=not operation_type.commutative,
+            )
+
+            for combination in combinations:
+                if not is_valid_combination(operation_type, combination):
+                    continue
+
+                expression = ExpressionWithNArgs(operation_type, ops=combination)
+                expressions.append(expression)
+
+    return expressions
+
+
+def get_combinations(
+    values: list[RawValue],
+    offset_value: RawValue,
+    length: int,
+    with_self: bool = True,
+    with_earlier: bool = True,
+) -> list[list[Expression]]:
+    if length == 0:
+        return []
+
+    value_combinations: list[list[Expression]] = []
+    start_reached = False
+
+    for current_value in values:
+        if current_value == offset_value:
+            start_reached = True
+
+        if not start_reached and not with_earlier:
+            continue
+
+        if current_value == offset_value and not with_self:
+            continue
+
+        if length == 1:
+            value_combinations.append([current_value])
+        else:
+            new_combinations = get_combinations(
+                values, current_value, length - 1, with_self, with_earlier
+            )
+
+            for new_combination in new_combinations:
+                entry: list[Expression] = [current_value]
+                entry.extend(new_combination)
+                value_combinations.append(entry)
+
+    return value_combinations
+
+
+def is_valid_combination(
```

Comment:
Planned to add operation-specific characterstics to determine further combinations resulting in an error (e.g., functions annotated with `VALUE_INCREASES` will fail if one operand has characteristic `MAX_VALUE` and at least one other has no characteristic like `ZERO_OR_NULL`.